### ISSUE TITLE
training accuracy dont go up by using MXnetR to build CNN network for CIFAR-10 dataset

### ISSUE THREAD

**Issue by rflai** [dmlc/mxnet#3527]
here is my mxnetR script.

rm(list=ls())
library(mxnet)
 
setwd("/home/mac/R_Projects/MXnet/CIFAR-10")
 
 train <- mx.io.ImageRecordIter(path.imglist = "/home/mac/R_Projects/MXnet/CIFAR-10/img2.txt",
                                                   path.imgrec = "/home/mac/R_Projects/MXnet/CIFAR-10/img2.rec",
                                                   data.shape = c(32,32,3),
                                                   batch.size = 100,
                                                   shuffle = TRUE
                                                    )
val <- mx.io.ImageRecordIter(path.imglist = "/home/mac/R_Projects/MXnet/CIFAR-10/val2.txt",
                                               path.imgrec = "/home/mac/R_Projects/MXnet/CIFAR-10/val2.rec",
                                               data.shape = c(32,32,3),
                                               batch.size = 100,
                                               shuffle = TRUE
                                                )

 ##- Define deep learning network, act_type can be relu, sigmoid, softrelu and tanh
 input_data <- mx.symbol.Variable(name = "data")
 
##- stage 1
conv1 <- mx.symbol.Convolution(data = input_data, kernel = c(5, 5), pad = c(2, 2), num_filter = 32, name = "conv1")
relu1 <- mx.symbol.Activation(data = conv1, act_type = "relu",  name = "relu1")
pool1 <- mx.symbol.Pooling(data = relu1, kernel = c(3, 3), stride = c(2, 2), pool_type = "max", name = "pool1")
 
##- stage 2
 conv2 <- mx.symbol.Convolution(data = pool1, kernel = c(5, 5), pad = c(2, 2), num_filter = 32, name = "conv2")
 relu2 <- mx.symbol.Activation(data = conv2, act_type = "relu", name = "relu2")
 pool2 <- mx.symbol.Pooling(data = relu2, kernel = c(3, 3), stride = c(2, 2), pool_type = "max", name = "pool2")
 
##- stage 3
 conv3 <- mx.symbol.Convolution(data = pool2, kernel = c(5, 5), pad = c(2, 2), num_filter = 64, name = "conv3")
 relu3 <- mx.symbol.Activation(data = conv3, act_type = "relu", name = "relu3")
 pool3 <- mx.symbol.Pooling(data = relu3, kernel = c(3, 3), stride = c(2, 2), pool_type = "max", name = "pool3")
 
##- stage 4
 flatten <- mx.symbol.Flatten(data=pool3)
 
##- stage 5
fc1 <- mx.symbol.FullyConnected(data = flatten, num_hidden = 64, name = "fc1")
relu4 <- mx.symbol.Activation(data = fc1, act_type = "relu", name = "relu4")

##- stage 6
fc2 <- mx.symbol.FullyConnected(data = relu4, num_hidden = 10, name = "fc2")
softmax <- mx.symbol.SoftmaxOutput(data = fc2, name = 'softmax')

mx.set.seed(2016)
device.gpu <- mx.gpu(0)
lr.rate <- c(0.0001, 0.001, 0.01, 0.1) 
for (n in 1: length(lr.rate)) {
cat("Process model train with learning rate ",lr.rate[n]," for IO data model is on going, please be patient......\n")
model <- mx.model.FeedForward.create(softmax,
                                                                 X=train,
                                                                 eval.data= val,
                                                                 ctx=device.gpu, 
                                                                 num.round=100,
                                                                 learning.rate=lr.rate[n],
                                                                 momentum=0.9,
                                                                 eval.metric=mx.metric.accuracy,
 ##                                                              initializer = mx.init.Xavier(factor_type='in', magnitude=2.34),
                                                                 wd=0.00001,
                                                                 batch.end.callback = mx.callback.log.train.metric(100)
                                                                  )
 }

the training accuracy is around 0.1. attached please find my img.txt file for prepare the img.rec data.

can anyone help on this?
[img.txt](https://github.com/dmlc/mxnet/files/529566/img.txt)
[val.txt](https://github.com/dmlc/mxnet/files/529569/val.txt)
[img2.txt](https://github.com/dmlc/mxnet/files/529571/img2.txt)
[val2.txt](https://github.com/dmlc/mxnet/files/529576/val2.txt)

[CIFAR_20161014V2.txt](https://github.com/dmlc/mxnet/files/529581/CIFAR_20161014V2.txt)

**Comment by rflai** [comment]
i also tried to load the CIFAR10 data into R array format and do normalization, then train the CNN by this model, seem situation happened.

here is my R script for loading CIFAR10 into R array format.

rm(list=ls())
library(mxnet)
library(imager)

setwd("/home/mac/R_Projects/MXnet/CIFAR-10")

Lists <- read.table("trainLabels_new.csv", header = TRUE, sep=",")
files <- list.files("./train/")

a <- array(0, c(32,32,3,20000))
train.y <- c()

for (i in 1:20000) {
  filename <- paste("",Lists$id[i],".png", sep="")
  if (length(intersect(filename, files)) == 1) {path <- paste("./train/",filename,"", sep="")
  test <- load.image(path)
  a[1:32,1:32,1,i] <- R(test)
  a[1:32,1:32,2,i] <- G(test)
  a[1:32,1:32,3,i] <- B(test)
  train.y[i] <- Lists$label[i]
  }
}

train.array <- a
train.sd.array <- a
mean.array <- array(0, c(32,32,3))
sd.array <- array(0, c(32,32,3))

for (j in 1:32){
  for (k in 1:32){
    for (l in 1:3){
      mean.array[j,k,l] <- mean(a[j,k,l,])
      sd.array[j,k,l] <- sd(a[j,k,l,])
    }
  }
}

for (m in 1:20000){
  train.array[,,,m] <- a[,,,m] - mean.array
  train.sd.array[,,,m] <- (a[,,,m] - mean.array)/sd.array
}

id <- sample(1:20000, 2000)

normal.val.array <- a[,,,id]
normal.val.y <- train.y[id]
normal.test.array <- a[,,,-id]
normal.test.y <- train.y[-id]

mean.val.array <- train.array[,,,id]
mean.val.y <- train.y[id]
mean.test.array <- train.array[,,,-id]
mean.test.y <- train.y[-id]

sd.val.array <- train.sd.array[,,,id]
sd.val.y <- train.y[id]
sd.test.array <- train.sd.array[,,,-id]
sd.test.y <- train.y[-id]

normal.train <- mx.io.arrayiter(data = normal.test.array, label = normal.test.y, batch.size = 100, shuffle = TRUE)
normal.val <- mx.io.arrayiter(data = normal.val.array, label = normal.val.y, batch.size = 100, shuffle = TRUE)

mean.train <- mx.io.arrayiter(data = mean.test.array, label = mean.test.y, batch.size = 100, shuffle = TRUE)
mean.val <- mx.io.arrayiter(data = mean.val.array, label = mean.val.y, batch.size = 100, shuffle = TRUE)

sd.train <- mx.io.arrayiter(data = sd.test.array, label = sd.test.y, batch.size = 100, shuffle = TRUE)
sd.val <- mx.io.arrayiter(data = sd.val.array, label = sd.val.y, batch.size = 100, shuffle = TRUE)

**Comment by rflai** [comment]
i did test the python version for train_cifar10.py, the accuracy can go up as normal, then i used the cifar-10 data downloaded and prepared by this python script (train.rec, test.rec) for the above R script, same situation happened, so it is not the data preparation caused this issue. no idea how to fix it, can anyone help?

**Comment by thirdwing** [comment]
I have add the cifar10 example for R in https://github.com/dmlc/mxnet/pull/3530

**Comment by thirdwing** [comment]
The result using lenet is as follow:
```
Rscript train_cifar10.R --cpu=TRUE --network=lenet 
Loading required package: mxnet
Loading required package: methods
Loading required package: argparse
Loading required package: proto
[1] "Network used: lenet"
[01:06:27] src/io/iter_image_recordio.cc:209: ImageRecordIOParser: data/cifar10/train.rec, use 1 threads for decoding..
[01:06:27] src/io/./iter_normalize.h:103: Load mean image from data/cifar10/mean.bin
[01:06:27] src/io/iter_image_recordio.cc:68: Loaded ImageList from data/cifar10/test.lst 10000 Image records
[01:06:27] src/io/iter_image_recordio.cc:209: ImageRecordIOParser: data/cifar10/test.rec, use 1 threads for decoding..
[01:06:27] src/io/./iter_normalize.h:103: Load mean image from data/cifar10/mean.bin
[1] "Computing with CPU"
Start training with 1 devices
Batch [50] Train-accuracy=0.25234375
Batch [100] Train-accuracy=0.284609375
Batch [150] Train-accuracy=0.302239583333333
Batch [200] Train-accuracy=0.3121875
Batch [250] Train-accuracy=0.32134375
Batch [300] Train-accuracy=0.328046875
Batch [350] Train-accuracy=0.332924107142857
[1] Train-accuracy=0.336237980769231
[1] Validation-accuracy=0.358880537974684
```

**Comment by rflai** [comment]
@thirdwing , thank you so much, i will go through it. by the way, i noticed some difference.
1. in the train-cifar10.py, the data_shape is (3,28,28), detail as following, but the cifar-10 pic i downloaded is (3,32,32). 
   def get_iterator(args, kv, data_shape=(3, 28, 28)):
    if '://' not in args.data_dir:
        _download(args.data_dir)

    train = mx.io.ImageRecordIter(
        path_imgrec = args.data_dir + "train.rec",
        mean_img    = args.data_dir + "mean.bin",
        data_shape  = data_shape,
        batch_size  = args.batch_size,
        rand_crop   = True,
        rand_mirror = True,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)

    val = mx.io.ImageRecordIter(
        path_imgrec = args.data_dir + "test.rec",
        mean_img    = args.data_dir + "mean.bin",
        rand_crop   = False,
        rand_mirror = False,
        data_shape  = data_shape,
        batch_size  = args.batch_size,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)
2.  the cifar data source difference. i opened the train.lst generated by the train_cifar10.py, found out the data seems are different with the one i downloaded. the filename i downloaded is just number, but the script downloaded with tag in the name.

5695    5       train/maltese_s_000167.png
55      4       train/cervus_elaphus_s_001601.png
3368    0       train/biplane_s_000041.png
6674    3       train/tabby_cat_s_002646.png
5975    1       train/estate_car_s_000247.png
7798    9       train/truck_s_000144.png


![10018](https://cloud.githubusercontent.com/assets/20311260/19407625/46857186-92da-11e6-9508-f62d78da0c36.png)
![167](https://cloud.githubusercontent.com/assets/20311260/19407629/6f745d1e-92da-11e6-87e0-40a534ea58d6.png)

**Comment by rflai** [comment]
by using the data download from this website, my R script works now. found out where the issue coming from. it is the data source. i downloaded the CIFAR-10 data from kaggle, and then use im2rec convert into the *.rec format. not sure which part did wrong, when training the model by this data, the accuracy just can't go up.
